### DISCUSSION TITLE
Trigger an 404 error from script setup

### DISCUSSION THREAD

**Discussion by ruslankonev** [nuxt/nuxt#17567] [Questions]
Which method we need to use in script setup to make a 404 error response page?

**Comment by klausbert** [comment]
Just make a default page named *.vue and write your 404 message there.

**Reply by ruslankonev** [reply]
@klausbert I mean how I can trigger error from script setup.

**Reply by ruslankonev** [reply]
`useAsyncData` returns an `{ error }` method, but using it crashes the app.

**Reply by wobsoriano** [reply]
> `useAsyncData` returns an `{ error }` method, but using it crashes the app.

Can you show your example?

**Reply by ruslankonev** [reply]
I was wrong, the returned `{error}` from `useAsyncData` is not a method, it is just a`ref` to report the error from the response.
I don’t know how to return an error if I didn’t receive a response from the server for the wrong parameter =).

**Reply by wobsoriano** [reply]
In Nuxt 2 we have this

```ts
try {
  // Some API request
} catch (err) {
  if (err.response.status === 404) {
    return this.$nuxt.error({ statusCode: 404, message: err.message })
  }
}
```

Not sure about nuxt 3

**Reply by ruslankonev** [reply]
> In Nuxt 2 we have this
> 
> ```ts
> try {
>   // Some API request
> } catch (err) {
>   if (err.response.status === 404) {
>     return this.$nuxt.error({ statusCode: 404, message: err.message })
>   }
> }
> ```
> 
> Not sure about nuxt 3

It is possible to use `$nuxt` inside script setup?

**Reply by wobsoriano** [reply]
You can access `error` inside `_legacyContext` but it's not yet supported.

```html
<script setup lang="ts">
const nuxtApp = useNuxtApp()

nuxtApp._legacyContext?.error({
  statusCode: 404,
  message: 'Some message'
})
</script>
```

You'll get `Accessing error is not yet supported in Nuxt 3.` in console

**Reply by ruslankonev** [reply]
Searching in source code I found something )

In `<script setup>` I can take `useNuxtApp` and write this 

```js
<script setup>
import { useNuxtApp } from '#app'

const { ssrContext } = useNuxtApp()
const { data: news, error } = await useAsyncData(route.params.id, () =>
  $fetch(`${config.baseURL}/api/news/${route.params.id}`)
)

if (error.value) {
  const error = new Error(`Page not found: ${ssrContext.url}`)
  error.statusCode = 404
  ssrContext.error = error
} else {
  news.value = {
    ...news.value,
    ...{
      content: wrapOutLinksInHTML(news.value.content),
    },
  }
}
</script>
```

Not sure, all of this correct, but I've got an 404 response =)

**Reply by Jos-de-Boss** [reply]
Not sure wether or not this is intended or not but I figured out a neat way you can do this as well.
You can create a custom Error class like this:

```js
// @/errors/NuxtNotFoundError.js
export default class NuxtNotFoundError extends Error {
  statusCode = 404;
}
```

Then in your pages or components you can simply use it with the [throwError utility][1] like this:
```vue
<!-- @/pages/mypage.vue -->
<script setup>
import NuxtNotFoundError from '@/errors/NuxtNotFoundError';

throwError(new NuxtNotFoundError('Resource not found'));
</script>
```

This way of doing things does still feel a little hacky. It would be nice to have a more official way of setting the response status.

[1]: https://v3.nuxtjs.org/guide/features/error-handling#throwerror

**Reply by Jos-de-Boss** [reply]
I did some further digging in the code. This is where Nitro inherits the `statusCode` property: https://github.com/unjs/nitro/blob/4fbeef9bfd5c8e064817865a31d61927addc5e2a/src/runtime/utils.ts#L63

Under the hood Nuxt3 uses a custom [H3Error][1] class from their H3 framework.

H3 also comes with a neat [createError][2] utility we can use here if we want.

Example:
```vue
<!-- @/pages/mypage.vue -->
<script setup>
import { createError } from 'h3';

throwError(
  createError({
    statusMessage: 'Resource not found',
    statusCode: 404,
  })
);
</script>
```

[1]: https://github.com/unjs/h3/blob/175ec7f88ab7a9fcaec698e952435fd2e35bcb0f/src/error.ts#L14-L18
[2]: https://www.jsdocs.io/package/h3#createError

**Comment by danielroe** [comment]
Issue to track: https://github.com/nuxt/framework/pull/4539
